File: src/main/java/com/bytetrend/sandbox/java/challenge/Find2ndSmallestNbr.java
package com.bytetrend.sandbox.java.challenge;

import java.util.Optional;

public class Find2ndSmallestNbr {
    public static Optional<Integer> find2ndSmallestNbr(int[] arr){
        if(arr == null || arr.length < 2)
            return Optional.empty();
        else{
            int frst = arr[0] < arr[1]?arr[0]:arr[1];
            int scnd = arr[0] > arr[1]?arr[0]:arr[1];
            for(int i = 2; i < arr.length; i++){
                if(arr[i] < scnd){
                    if(arr[i]<frst) {
                        scnd = frst;
                        frst = arr[i];
                    }else{
                        scnd = arr[i];
                    }
                }
            }
            return Optional.of(scnd);
        }
    }
    public static void result(Optional<Integer> scnd){
        if(scnd.isPresent())
            System.out.println("second smallest is: "+scnd.get());
        else
            System.out.println("Array did not contain 2 or more numbers");
    }
    public static void main(String[] args){
        int[] arr = new int[]{9, 3, 0, 4, 5, 8, 7, 6, 1, 2};
        result(find2ndSmallestNbr(arr));
        result(find2ndSmallestNbr(null));
    }
}

File: src/main/scala/com/bytetrend/sandbox/scala/challenge/Find2ndSmallestNbr.scala
package com.bytetrend.sandbox.scala.challenge

object Find2ndSmallestNbr {

  def main(args: Array[String]) {
    def result(x:Option[Int])= x match {
      case Some(x) => println(s"second smallest is: $x")
      case _ => println("Array did not contain 2 or more numbers")
    }
    val arr = Array[Int](9, 3, 0, 4, 5, 8, 7, 6, 1, 2)
    result(find2ndSmallestNbr(arr) )
    result(find2ndSmallestNbr(Array[Int]()))
  }

  def find2ndSmallestNbr(arr: Array[Int]): Option[Int] = {
    arr match {
      case a: Array[Int] if a.length > 1 => {
        //(first,second)
        Some(arr.tail.tail.foldLeft(if (arr.head < arr.tail.head) (arr.head, arr.tail.head) else (arr.tail.head, arr.head)) {
          case ((frst, scnd), n) => if (n < scnd) {
            if (n < frst) (n, frst) else (frst, n)
          } else (frst, scnd)
        }._2)
      }
      case _ => None
    }
  }
}
